src/data/equiv/fin.lean

```diff
@@ -149,6 +169,63 @@ def sum_fin_sum_equiv : fin m ⊕ fin n ≃ fin (m + n) :=
     { dsimp, rw [dif_neg H], simp [fin.ext_iff, nat.add_sub_of_le (le_of_not_gt H)] }
   end }
 
+/-- The equivalence between `fin (m + n)` and `fin (n + m)` which rotates by `n`. -/
+def fin_add_flip : fin (m + n) ≃ fin (n + m) :=
+  (sum_fin_sum_equiv.symm.trans (equiv.sum_comm _ _)).trans sum_fin_sum_equiv
+
+@[simp] lemma fin_add_flip_apply_left {k : ℕ} (h : k < m) :
+  fin_add_flip (⟨k, nat.lt_add_right k m n h⟩ : fin (m + n)) = ⟨n + k, add_lt_add_left h n⟩ :=
```

Comment:
```suggestion
@[simp] lemma fin_add_flip_apply_left {k : ℕ} (h : k < m) (hk : k < m + n := nat.lt_add_right k m n h)
  (hnk : n + k < m + n := add_lt_add_left h n) :
  fin_add_flip (⟨k, nk⟩ : fin (m + n)) = ⟨n + k, hnk⟩ :=
```
otherwise this only matches proofs constructed exactly that way

Reply:
Oh! I thought no matching ever happened on proofs, in either `simp` or `rw`.

Reply:
I used to think that too, but I think @gebner and a couple unprincipled tests convinced me that matching does happen. Maybe it only matters for `rw`?

Reply:
Let me first explain what the problem is.  When simp (or rw, it doesn't matter here) tries to rewrite using this lemma, it clearly needs to find a proof of `k < m`.  Like any other argument, the value for `h` is found by matching.  If the term that you want to rewrite literally contains `fin_add_flip ⟨k, nat.lt_add_right k m n hhhhhh⟩`, then matching will succeed and use `hhhhhh` as the value for the `h` argument.  But if you want to rewrite `fin_add_flip ⟨k, hhh⟩`, then matching won't work: it only sees `hhh : k < m + n` but expects a `nat.lt_add_right k m n ???`.  (There's also no way it can work as we need a proof of `k < m`, but only have `k < m + n`.)

Apparently, in this case matching will not fail but produce a new goal for `k < m`.  So the optional arguments are unnecessary.

Usually I would suggest to rewrite this lemma as follows, but alas this won't work here:
```lean
@[simp] lemma fin_add_flip_apply_left {k : ℕ} (hk : k < m + n) :
  fin_add_flip (⟨k, hk⟩ : fin (m + n)) = ⟨n + k, nat.lt_of_add_right hk⟩ :=
```

I'm not convinced this is a good simp lemma.  It requires simp to prove `k < m`, which it probably can't in most cases.